Update release instructions (#2896)
grafana/agent: docs/developer/releasing.md

## Performing Releases

The lifetime of a release branch is shepherded by a single person, picked from a
rota of project maintainers and contributors. The release shepherd is
responsible for managing all releases that compose the release branch.

For a **new release branch**, the release shepherd will:

1. Gather consensus on which commit should be used as a base for the release
   branch.

2. Create and push the release branch from the selected base commit. 

   The name of the release branch should be the name of the stable version we intend 
   to release, e.g. `release-v0.31`. Release branches do not contain the `-rc.N` release 
   candidate suffix, i.e. there is no `release-v0.31-rc.0`.

3. Create a PR to cherry-pick additional commits into the release branch as
   needed.

> 4. Create a PR to [update code](#updating-code) using the `-rc.N` release

Reviewer: We should note that branches used for PRs should have a name that doesn't start with `release-`, otherwise branch protection rules apply to it. Alternatively, branches used for PRs can come from forks.

Author: Thank you, I'll add it like this:

```
4. Create a PR to [update code](#updating-code) using the `-rc.N` release
   candidate version.

   NOTE: PRs should have a name that doesn't start with `release-`, otherwise branch 
   protection rules apply to it. Alternatively, branches used for PRs can come from forks.
```

   candidate version.

> 5. Tag the release candidate using the tag naming convention `vX.Y.0-rc.N`.

Reviewer: I'm worried that the instructions here might be a little confusing. We should publish the release candidate too, but the instructions don't mention it. 

Maybe there should be a dedicated section for post-publishing steps, where:

For a release candidate:

* Publish the release candidate 
* Announce the release candidate in the public Slack channel 

For a stable release:

* Publish the release candidate 
* Ensure Homebrew gets updated 
* Ensure documentation got published 
* Announce the stable release in the public Slack channel

Author: True, it does get confusing. I was trying to think how to implement the  post-publishing steps suggestion, but couldn't think of a super clean way.

For now I split the entire "candidate" and the "stable" steps into two different lists. I feel like this makes sense, since they are usually done on two different days? That way we start from point 1 on the second day.

## Performing Releases

The lifetime of a release branch is shepherded by a single person, picked from a
rota of project maintainers and contributors. The release shepherd is
responsible for managing all releases that compose the release branch.

> For a **new release candidate**, the release shepherd will:

Reviewer: Personally, I preferred the older instructions, because it talked about a release branch as a unit: a stream of releases for the same major/minor version tuple. 

I'm concerned saying "for a new release candidate" may be confusing, because the same minor release branch may contain any number of release candidates, depending on how many bugs are discovered. 

Maybe we need a new term here to refer to these branches? "Release streams?"

Author: My reasoning for splitting this into sections of "new release candidate", "stable release" and "patch releases" is that people will likely do them at three separate times. When someone finishes with the "new release candidate" section they are supposed to pause for a few days and then resume from the next section on "stable release". It makes it a little easier to find where to resume, instead of looking for the correct point in a list.

If we could make the headings something like "Step 1: New release candidate", "Step 2: Stable release" and "Step 3: Patch releases", maybe that'd make it more clear that those are different steps of the same continuous process?

In my case, I suppose the "unit" is the Github release and the outcome of each of the three sections is a "release" as in a "Github release". I agree there is value in thinking of the release branch itself as a unit though. Maybe instead of "Performing Releases", this whole section should be called "Release branch lifecycle"?

1. Gather consensus on which commit should be used as a base for the release
   branch.

2. Create and push the release branch from the selected base commit:

>    * The name of the release branch should be the name of the stable version we intend 
>    to release, e.g. `release-v0.31`.
>    * Release branches do not contain the `-rc.N` release candidate suffix. I.e. there 
>    is no branch called `release-v0.31-rc.0`.
>    * Release branches do not contain the patch version. I.e. there is no branch called 
>    `release-v0.31.0` or `release-v0.31.4`.

Reviewer: The Google documentation style guide recommends against using e.g. and i.e. since people tend to mix them up.

```suggestion
   * The name of the release branch should be the name of the stable version we intend 
   to release, such as `release-v0.31`.
   * Release branches do not contain the `-rc.N` release candidate suffix. This means
     there is no branch called `release-v0.31-rc.0`.
   * Release branches do not contain the patch version. This means there is no 
     branch called `release-v0.31.0` or `release-v0.31.4`.
```

3. Create a PR to cherry-pick additional commits into the release branch as
   needed.

4. Create a PR to [update code](#updating-code) on `main` using the `-rc.N` release
   candidate version.

5. Create a PR to cherry pick the update from step 4 to the release branch.

> 5. [Create a tag](#create-a-tag).

Reviewer: Your numbering here drifts (there's two 5s).

   fixes are available, cherry-picking the fixes into the release branch and
   starting a new release candidate.

> 7. Create a PR to [update code](#updating-code) again - this time using the stable

Reviewer: ```suggestion
7. Create a new PR to [update code](#updating-code) using the _stable_
```

   release version. 

8. Create the stable release tag.

> 9. Make sure [Homebrew is updated](#updating-homebrew).

Reviewer: The order here seems wrong. Updating Homebrew can only happen after the release is published.

If no new issues were found during the environment testing, the shepherd can continue with 
creating a **stable release**:

1. Create a new PR to [update code](#updating-code) using the _stable_
   release version (i.e. without `-rc.N`).

2. Create the stable release tag.

3. Force-push the `release` branch to point at the stable release tag. This
   branch is used to externally reference files in the repository for a stable
   release, e.g. the "[latest](https://grafana.com/docs/agent/latest/)" documentation.

4. Make sure the documentation is updated:
    1. A new version of the documentation should be visible on the 
    [version switcher](https://grafana.com/docs/versions/?project=/docs/agent/).
    2. The "[latest](https://grafana.com/docs/agent/latest/)" one should be identical to the latest version.
    3. Check the "Upgrade guide" - the latest changes listed in it should be assigned to the latest version.

5. [Publish the release on Github](#publishing-the-release)

6. [Update Agent Operator Helm chart version](#update-agent-operator-helm-chart-version)

7. Make sure [Homebrew is updated](#updating-homebrew).

8. Announce the release on community Slack channel.

> For **patch releases**, the release shepherd will:

Author: Hi @rfratto, do you think that the steps below are good enough to describe patch releases? If you agree, I could update the md file.

#### For **patch releases**, the release shepherd will:
1. Merge the fixes we need for the patch to `main`.
2. Raise a PR to `main` to do two things:
    1. Add a new section to the changelog for the patch release. **Move** the fixes listed in the changelog via step 1 from the "unreleased" section to the section for the new patch release.
    2. Update code so that `main` is referencing the new patch version in all files.
4. After the PR in step 2 is merged, cherry pick the commit from step 2 into the release branch.
5. Create a new tag for the patch release.

The above instructions have a few implications:
- When creating patch releases, there is no need for a release candidate.
- Changes made to patch releases are not listed in the changelog for the next stable version.
- The tag of the patch release does yet not exist at the time when we update code for the main branch to reference it.

NOTE: When patching an old release branch, we would do some things differently:
- `main` would not be updated to reference the patch version
- The change applied to the old release branch would have to be listed in two sections in the changelog:
        1. The patch release's section
        2. The "unreleased" section

Reviewer: These look good 👍

For **patch releases**, the release shepherd will:

1. Merge the fixes we need for the patch to `main`.
   
2. Cherry-pick the fixes from `main` into the release branch.

3. Follow the same instructions as the ones for a stable release, starting from step 1.

Some notes on patch releases:

- When creating patch releases, there is no need for a release candidate.

- Changes made to patch releases are not listed in the CHANGELOG for the next stable version.

- The tag of the patch release does yet not exist at the time when we update code for the 
main branch to reference it.

### Updating code

The codebase must be updated to reference the upcoming release tag whenever a new release is being made.

NOTE: Any change done for a release should be done in `main` first and then moved to the release branch.

NOTE: Branches used for PRs should have a name that doesn't start with `release-`, 
otherwise branch protection rules apply to it. Alternatively, branches used for PRs 
can come from forks.

#### Update the CHANGELOG

When creating a release **candidate** (i.e. an `-rc.N` version):

* Add a new section in the CHANGELOG for the new release candidate and include 

> today's date. E.g. "v0.31.0-rc.0 (2023-01-26)".

Reviewer: ```suggestion
the date of the release, for example, "v0.31.0-rc.0 (2023-01-26)".
```

3. Create the patch release tag.

4. [Publish the release](#publishing-the-release)

### Updating code

The codebase must be updated whenever a new release is being made to reference the upcoming release tag.

#### Update the changelog

> NOTE: Any time CHANGELOG.md is updated for a release, it should first be done

Reviewer: We should move this warning and re-word it a little. Any change done for a release should be done in main first and then moved to the release branch.

Author: Good point! I'll move it to the start of the "Updating code" section, like this:

```
### Updating code

The codebase must be updated whenever a new release is being made to reference the upcoming release tag.

NOTE: Any change done for a release should be done in `main` first and then moved to the release branch.
```

via PR to the release branch, and then by a second PR to main.

When creating a release **candidate** (an `-rc.N` version):

* Add a new section in the changelog for the new release candidate and include 
today's date. E.g. "v0.31.0-rc.0 (2023-01-26)".
* All items form the "unreleased" section will move to a new section for the upcoming release version.
* See [here](https://github.com/grafana/agent/pull/2838/files) for an example pull request.

> * Sanity check that the changelog doesn't contain features which don't belong to it. Sometimes features get added to the wrong version in the changelog due to bad merges.

Reviewer: ```suggestion
* Ensure that the changelog doesn't contain features which don't belong to it. Sometimes features get added to the wrong version in the changelog due to bad merges.
```

When creating a **stable** release:

* Replace the version from the release candidate section to be the stable release instead 
of the `-rc.N` version.
* We do not leave release candidates in the changelog once there is a stable version for them.
* Make sure to also update the **date** in the title.
* See [here](https://github.com/grafana/agent/pull/2873/files) for an example pull request.

> #### Search and replace the old version with the new one

Reviewer: Can we add a notice here to not touch the `operations/helm` directory? That should be updated independently from releases for now.

Author: Will do, thanks!

Go through the entire repository and find references to the previous release
  version, updating them to reference the new version **where necessary**.

NOTE: There are files such as 
"[pkg/operator/defaults.go](https://github.com/grafana/agent/blob/main/pkg/operator/defaults.go)", 
where the version sometimes should not be replaced but added to a list of versions.
At the time of this writing, `defaults.go` is the only such file.

#### Update defaults.go

Add the new version to the "[pkg/operator/defaults.go](https://github.com/grafana/agent/blob/main/pkg/operator/defaults.go)" file. If there is a release candidate (`-rc.N`) version, remove it.

#### Update manifests and dashboards

Run `make generate-manifests` and `make generate-dashboards` to update

> manifests in case they are stale.

Reviewer: Ideally we remove this from the release instructions and instead run `make generate-manifests generate-dashboards` in CI and fail a PR which causes them to be stale.

Author: If I understand you correctly, what these two `make` commands do is the same as what we would do by search and replace of the version through the codebase? Hence we can do search and replace for now as our release process, and use the `make` commands just to double check that we replaced everything correctly?

I am not too sure how to make sure those `make` commands completed successfully. Maybe the best way would be to have the makefile return an error code which causes Drone to fail the CI? Here I am guessing that Drone would fail when one of the `commands` in the docker pipeline returns a non-zero code.

But then, how would the makefile know that the command failed? I ran a `tk show` command just to see what happens when `tk` exits with an error, and the return code was 0. A 0 normally indicates success. Maybe there is a way other than error codes to check for failures?

If you want to, I could move this into a separate issue and could mention that we should remove the steps from the documentation when this is added to the CI?

Reviewer: I'm suggesting that we have a CI job to make sure that the manifests are never stale. We do this for Helm chart tests: https://github.com/grafana/agent/blob/9c068ac3993983b1e3160af6e9af117fdda35f64/.github/workflows/helm-test.yml#L34-L41

That way, there would be no reason to include this step in the release step instructions because manifests would not be stale at release time.

Author: Ok, thank you, I'll delete the instructions from here and will make sure to update the CI on the other [PR](https://github.com/grafana/agent/pull/2966).

### Merge freezes

Release shepherds may request a merge freeze to main for any reason during the
release process.

> ### Publishing the Release

Reviewer: Can we add a warning somewhere to note that you should never force push a tag after the publish button is pressed for a release?

   CHANGELOG. Add a link to the CHANGELOG, noting that the full list of changes
   can be found there. Refer to other releases for help with formatting this.

3. Tick the appropriate boxes at the bottom of the release page:

      * For release candidates:

         1. Tick the checkbox to "set as pre-release".

      * For stable releases:

         1. Tick the checkbox to "set as the latest release".

         2. Tick the check box to "add a discussion" under the category for "announcements".

4. Optionally, have other team members review the release draft if you wish
   to feel more comfortable with it.

5. Publish the release!

> NOTE: Release candidates should be retained on the 
> [GitHub releases page](https://github.com/grafana/agent/releases).
> Please do not remove them.

Reviewer: ```suggestion
> **NOTE**: Release candidates should be retained on the 
[GitHub releases page](https://github.com/grafana/agent/releases).
Please do not remove them.
```

2. Update references of agent-operator app version in helm-charts pointing to release version and bump helm chart version.
You can use this [pull-request](https://github.com/grafana/helm-charts/pull/1831) as reference.

### Updating Homebrew
[Homebrew](https://brew.sh/) is a package manager for MacOS. It installs binaries 
(aka "[bottles](https://docs.brew.sh/Bottles)") from either the main Homebrew repository 
or a third-party one (aka a "[tap](https://docs.brew.sh/Taps)").

For every Agent release we need to update two Brew repositories:
* [homebrew/homebrew-core](https://github.com/Homebrew/homebrew-core) is the main one which 
Brew installations use by default

> * [grafana/homebrew-grafana](https://github.com/grafana/homebrew-grafana) is a "tap" - 

Reviewer: Rather than documenting our custom tap, should we follow in Mimir's footsteps and deprecate our own Formula in favor of the official Homebrew core one? See grafana/homebrew-grafana#41.

Author: The custom tap is still on 31.0, so it's been effectively deprecated...I'll remove it from the instructions for now and in the unlikely event that we are unable to decommission it we could add it back.
